### PR TITLE
[libc] abs is a const function

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

**Comment by jhuber6** [comment]
@michaelrj-google Do we have a formalized way to specify attributes? My gut says it should go in `libc/spec/stdc.td` but I remember hearing there was some desire to move away from tablegen?

**Comment by AtariDreams** [comment]
> @michaelrj-google Do we have a formalized way to specify attributes? My gut says it should go in `libc/spec/stdc.td` but I remember hearing there was some desire to move away from tablegen?

I have no idea how to do it in tablegen

**Comment by jhuber6** [comment]
As it stands, this won't be exported. The header modified is simply the internal header. The installed ones go through a tool called `lib-hdrgen` which uses LLVM TableGen to spit the definitions out into where `%%public_api()` is.

File: libc/src/stdlib/abs.h [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 namespace LIBC_NAMESPACE {
 
-int abs(int n);
+__attribute__((__const__)) int abs(int n);
```
Comment (SchrodingerZhu) [inline libc/src/stdlib/abs.h:14]: `[[gnu::const]]`

**Review by SchrodingerZhu — CHANGES_REQUESTED** [review]

**Comment by jhuber6** [comment]
> I have no idea how to do it in tablegen

I was asking Michael, because I recall that this is an existing issue that the infrastructure hasn't solved yet. I had problems earlier where it conflicted with a lot of GNU definitions that expect things like `pure` or similar.

**Comment by SchrodingerZhu** [comment]
We have internal versions of math kernels `LIBC_NAMESPACE`. I think it is legitimate to add additional attributes to these internal versions of these functions if found helpful. Addition non-standard attributes to exposed version as specified in spec sometimes will lead problems.

**Review by SchrodingerZhu — APPROVED** [review]
This approval is to dismiss previous "Request changes". I don't know how to dismiss a "Request changes" without approving the PR.

**Comment by nickdesaulniers** [comment]
Isn't https://github.com/llvm/llvm-project/pull/79891 trying to add generalized support for attributes?

**Comment by jhuber6** [comment]
> Isn't #79891 trying to add generalized support for attributes?

I think the attributes on the installed headers and the internal headers are separate.

**Pull request closed without merging**